Don't let a blocked maintenance operation inhibit remaining maintenance queue [run-systemtest]

vekterli opened the pull request:
@geirst please review
@toregge FYI

The old maintenance scheduler behavior is to only remove a bucket from the
priority DB if its maintenance operation was successfully started. Failing
to start an operation could happen from both max pending throttling as well
as operation/bucket-specific blocking behavior. Since the scheduler would
encounter the same bucket as the one previously blocked upon its next tick
invocation, a single blocked bucket would run the risk of head-of-line
stalling the rest of the remaining maintenance queue (assuming the ongoing
DB scan did not encounter any higher priority buckets).

This commit changes the following aspects of maintenance scheduling:
 * Always clear entries from the priority DB before trying to start an
   operation. A blocked operation will be retried the next time the
   regular bucket DB scan encounters the bucket.
 * Avoid trying to start (and clear) inherently doomed operations by
   _not_ trying to schedule any operations if it would be blocked due
   to too many pending maintenance operations anyway. Introduces a
   new `PendingWindowChecker` interface for this purpose.
 * Explicitly inhibit all maintenance scheduling if a pending cluster
   state is present. Operations are already _implicitly_ blocked from
   starting if there's a pending cluster state, but this would cause
   the priority DB from being pointlessly cleared.

vekterli committed:
Don't let a blocked maintenance operation inhibit remaining maintenance queue

The old maintenance scheduler behavior is to only remove a bucket from the
priority DB if its maintenance operation was successfully started. Failing
to start an operation could happen from both max pending throttling as well
as operation/bucket-specific blocking behavior. Since the scheduler would
encounter the same bucket as the one previously blocked upon its next tick
invocation, a single blocked bucket would run the risk of head-of-line
stalling the rest of the remaining maintenance queue (assuming the ongoing
DB scan did not encounter any higher priority buckets).

This commit changes the following aspects of maintenance scheduling:
 * Always clear entries from the priority DB before trying to start an
   operation. A blocked operation will be retried the next time the
   regular bucket DB scan encounters the bucket.
 * Avoid trying to start (and clear) inherently doomed operations by
   _not_ trying to schedule any operations if it would be blocked due
   to too many pending maintenance operations anyway. Introduces a
   new `PendingWindowChecker` interface for this purpose.
 * Explicitly inhibit all maintenance scheduling if a pending cluster
   state is present. Operations are already _implicitly_ blocked from
   starting if there's a pending cluster state, but this would cause
   the priority DB from being pointlessly cleared.

geirst commented on storage/src/vespa/storage/distributor/maintenance/pending_window_checker.h:
```diff
@@ -0,0 +1,14 @@
+
+namespace storage::distributor {
+
+class PendingWindowChecker {
```
Consider adding comment.

vekterli replied:
Will do this in a follow-up

geirst reviewed (approved):
👍  @toregge please review as well

vekterli committed:
Make implicit bucket priority DB clearing on scheduling configurable

vekterli committed:
Use blocking scheduling semantics for bucket activation maintenance

We consider bucket maintenance so latency critical that we'll prefer
to stall scheduling of subsequent buckets instead of risking having
to re-scan the DB to encounter the bucket again.

geirst reviewed (approved):
👍
